File: Data Structures & Algorithms/word-break/submission-1.cpp
class Solution {
public:
    unordered_map<int, bool> memo;

    bool wordBreak(string s, vector<string>& wordDict) {
        memo[s.length()] = true;
        return dfs(s, wordDict, 0);
    }

    bool dfs(string& s, vector<string> &wordDict, int i) {
        if (memo.find(i) != memo.end()) {
            return memo[i];
        }

        for (const string& word : wordDict) {
            if (word.length() + i <= s.length() && s.substr(i, word.length()) == word) {
                if (dfs(s, wordDict, i + word.length())) {
                    memo[i] = true;
                    return true;
                }
            }
        }
        
        memo[i] = false;
        return false;
    }
};

File: Data Structures & Algorithms/count-subsequences/submission-1.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m, n = len(s), len(t)
        dp = [0] * (n + 1)
        nextDp = [0] * (n + 1)
        dp[n] = nextDp[n] = 1


        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                nextDp[j] = dp[j]
                if s[i] == t[j]:
                    nextDp[j] += dp[j + 1]
            dp = nextDp[:]
        
        return dp[0]

        # memo = {}
        
        # def dfs(i, j):
        #     if j == len(t):
        #         return 1
        #     if i >= len(s):
        #         return 0
            
        #     if (i, j) in memo:
        #         return memo[(i, j)]

        #     res = 0
        #     if s[i] == t[j]:
        #         res = dfs(i + 1, j + 1)
        #     res += dfs(i + 1, j)
        #     memo[(i, j)] = res
        #     return res

        # return dfs(0, 0)